File: presenterRx2/src/main/java/me/ameriod/lib/mvp/presenter/rx2/IObservableSchedulerRx2.java
package me.ameriod.lib.mvp.presenter.rx2;

import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.ObservableTransformer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.annotations.NonNull;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Action;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;
import io.reactivex.schedulers.TestScheduler;

public interface IObservableSchedulerRx2 {

    <T> ObservableTransformer<T, T> schedule();


    IObservableSchedulerRx2 SUBSCRIBE_IO_OBSERVE_ANDROID_MAIN = new IObservableSchedulerRx2() {
        @Override
        public <T> ObservableTransformer<T, T> schedule() {
            return new ObservableTransformer<T, T>() {
                @Override
                public ObservableSource<T> apply(Observable<T> upstream) {
                    return upstream
                            .subscribeOn(Schedulers.io())
                            .observeOn(AndroidSchedulers.mainThread());
                }
            };
        }
    };

    IObservableSchedulerRx2 SUBSCRIBE_COMPUTATION_OBSERVE_ANDROID_MAIN = new IObservableSchedulerRx2() {
        @Override
        public <T> ObservableTransformer<T, T> schedule() {
            return new ObservableTransformer<T, T>() {
                @Override
                public ObservableSource<T> apply(Observable<T> upstream) {
                    return upstream
                            .subscribeOn(Schedulers.computation())
                            .observeOn(AndroidSchedulers.mainThread());
                }
            };
        }
    };

    IObservableSchedulerRx2 TRAMPOLINE = new IObservableSchedulerRx2() {
        @Override
        public <T> ObservableTransformer<T, T> schedule() {
            return new ObservableTransformer<T, T>() {
                @Override
                public ObservableSource<T> apply(Observable<T> upstream) {
                    TestScheduler scheduler = new TestScheduler();
                    return upstream.subscribeOn(Schedulers.trampoline())
                            .observeOn(Schedulers.trampoline());
                }
            };
        }
    };

    IObservableSchedulerRx2 COMPUTATION = new IObservableSchedulerRx2() {
        @Override
        public <T> ObservableTransformer<T, T> schedule() {
            return new ObservableTransformer<T, T>() {
                @Override
                public ObservableSource<T> apply(Observable<T> upstream) {
                    return upstream.subscribeOn(Schedulers.computation())
                            .observeOn(Schedulers.computation());
                }
            };
        }
    };

    IObservableSchedulerRx2 IO = new IObservableSchedulerRx2() {
        @Override
        public <T> ObservableTransformer<T, T> schedule() {
            return new ObservableTransformer<T, T>() {
                @Override
                public ObservableSource<T> apply(Observable<T> upstream) {
                    return upstream.subscribeOn(Schedulers.io())
                            .observeOn(Schedulers.io());
                }
            };
        }
    };

}

File: presenterRx2/src/main/java/me/ameriod/lib/mvp/presenter/rx2/IObservableSchedulerRx2.kt
package me.ameriod.lib.mvp.presenter.rx2

import io.reactivex.ObservableTransformer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import io.reactivex.schedulers.TestScheduler

interface IObservableSchedulerRx2 {

    fun <T> schedule(): ObservableTransformer<T, T>

    companion object {

        val SUBSCRIBE_IO_OBSERVE_ANDROID_MAIN: IObservableSchedulerRx2 = object : IObservableSchedulerRx2 {
            override fun <T> schedule(): ObservableTransformer<T, T> {
                return ObservableTransformer { upstream ->
                    upstream
                            .subscribeOn(Schedulers.io())
                            .observeOn(AndroidSchedulers.mainThread())
                }
            }
        }

        val SUBSCRIBE_COMPUTATION_OBSERVE_ANDROID_MAIN: IObservableSchedulerRx2 = object : IObservableSchedulerRx2 {
            override fun <T> schedule(): ObservableTransformer<T, T> {
                return ObservableTransformer { upstream ->
                    upstream
                            .subscribeOn(Schedulers.computation())
                            .observeOn(AndroidSchedulers.mainThread())
                }
            }
        }

        val TRAMPOLINE: IObservableSchedulerRx2 = object : IObservableSchedulerRx2 {
            override fun <T> schedule(): ObservableTransformer<T, T> {
                return ObservableTransformer { upstream ->
                    val scheduler = TestScheduler()
                    upstream.subscribeOn(Schedulers.trampoline())
                            .observeOn(Schedulers.trampoline())
                }
            }
        }

        val COMPUTATION: IObservableSchedulerRx2 = object : IObservableSchedulerRx2 {
            override fun <T> schedule(): ObservableTransformer<T, T> {
                return ObservableTransformer { upstream ->
                    upstream.subscribeOn(Schedulers.computation())
                            .observeOn(Schedulers.computation())
                }
            }
        }

        val IO: IObservableSchedulerRx2 = object : IObservableSchedulerRx2 {
            override fun <T> schedule(): ObservableTransformer<T, T> {
                return ObservableTransformer { upstream ->
                    upstream.subscribeOn(Schedulers.io())
                            .observeOn(Schedulers.io())
                }
            }
        }
    }

}
